### DISCUSSION TITLE
Selecting and entering value in Stripe Elements (card details) for payment testing

### DISCUSSION THREAD

**Discussion by arsalanator** [webdriverio/webdriverio#7431] [Q&A]
I am trying to get to Stripe Elements with webdriverIO so that I can enter card details but failing to get to that element.
Below is how I am trying to reach it
`
<pre class="file-source no-highlight hljs" id="c446e0be3cd44bc5a4eb7e0aabb8ef5b" data-ln-prefix="accessingelement">

const Page = require('./page');
--
  |  
  | class SettingsPage extends Page {
  |  
  | get stripeCardNumberInput () {
  | let outerDiv = $('#stripeCheckout')
  | return outerDiv.$('.InputElement')
  | } // $('[name="cardnumber"]')
  | get stripeCardExpiryInput () { return  } // $('[name="cardExpiry"]')
  | get stripeCardCVCInput () { return  } // $('[name="cvc"]')
  | get subscribeButton () { return $('#subscribeButton') }
  |  
  | async waitForStripeCheckout () {
  | $('#stripeCheckout').waitForClickable({ timeout:5000 })
  | // $('#stripeCheckout').waitForExist({ timeout:5000 })
  | }
  |  
  | async enterCardDetailsAndSubmit (cardNumber, cardExpiry, cvc) {
  | await this.stripeCardNumberInput.setValue(cardNumber);
  | await this.stripeCardExpiryInput.setValue(cardExpiry);
  | await this.stripeCardCVCInput.setValue(cvc);
  | await this.subscribeButton.click();
  | }
  | }

</pre>`

and below is the error log

`
<pre class="file-source no-highlight hljs" id="fd7f5e4a55bc4563ab16af9f41e8a8ed" data-ln-prefix="error-log">

[0-0] 2021-09-15T14:19:41.229Z INFO webdriver: DATA { using: 'css selector', value: '.InputElement' }
--
  | [0-0] 2021-09-15T14:19:41.243Z INFO webdriver: RESULT []
  | [0-0] 2021-09-15T14:19:41.729Z INFO webdriver: COMMAND findElementsFromElement("25a453ab-2491-40ff-b66e-73ec10a335a2", "css selector", ".InputElement")
  | [0-0] 2021-09-15T14:19:41.730Z INFO webdriver: [POST] http://localhost:9515/session/f198108fd6e3ab28dfd5e5b5d2e94f21/element/25a453ab-2491-40ff-b66e-73ec10a335a2/elements
  | [0-0] 2021-09-15T14:19:41.730Z INFO webdriver: DATA { using: 'css selector', value: '.InputElement' }
  | [0-0] 2021-09-15T14:19:41.743Z INFO webdriver: RESULT []
  | [0-0] 2021-09-15T14:19:42.231Z INFO webdriver: COMMAND takeScreenshot()
  | [0-0] 2021-09-15T14:19:42.232Z INFO webdriver: [GET] http://localhost:9515/session/f198108fd6e3ab28dfd5e5b5d2e94f21/screenshot
  | [0-0] 2021-09-15T14:19:42.433Z INFO webdriver: RESULT iVBORw0KGgoAAAANSUhEUgAABBIAAAJOCAYAAAAKxCRwAAAAAXNSR0IArs4c6...
  | [0-0] Error in "Signup and Get Subscription.should signup, login with credentials, and get stripe monthly subscription"
  | Error: Can't call setValue on element with selector ".InputElement" because element wasn't found
  | at Object.implicitWait [as default] (/home/arsalan/Work/Job/utak/e2eTestingForUtakWeb/node_modules/webdriverio/build/utils/implicitWait.js:34:19)
  | at async Element.elementErrorHandlerCallbackFn (/home/arsalan/Work/Job/utak/e2eTestingForUtakWeb/node_modules/webdriverio/build/middlewares.js:20:29)
  | at async Element.wrapCommandFn (/home/arsalan/Work/Job/utak/e2eTestingForUtakWeb/node_modules/@wdio/utils/build/shim.js:106:29)
  | at async SettingsPage.enterCardDetailsAndSubmit (/home/arsalan/Work/Job/utak/e2eTestingForUtakWeb/test/pageobjects/settings.page.js:44:9)
  | at async Context.<anonymous> (/home/arsalan/Work/Job/utak/e2eTestingForUtakWeb/test/specs/assignSubscriptionToNewUser.js:27:13)
  | [0-0] 2021-09-15T14:19:52.498Z INFO webdriver: COMMAND deleteSession()
  | [0-0] 2021-09-15T14:19:52.498Z INFO webdriver: [DELETE] http://localhost:9515/session/f198108fd6e3ab28dfd5e5b5d2e94f21
  | [0-0] FAILED in chrome - /test/specs/assignSubscriptionToNewUser.js
  | 2021-09-15T14:19:52.667Z INFO @wdio/cli:launcher: Run onComplete hook
  |  
  | Spec Files:	 0 passed, 1 failed, 1 total (100% completed) in 00:01:01
  |  
  | 2021-09-15T14:19:52.668Z INFO @wdio/local-runner: Shutting down spawned worker
  | 2021-09-15T14:19:52.920Z INFO @wdio/local-runner: Waiting for 0 to shut down gracefully
  | 2021-09-15T14:19:52.920Z INFO @wdio/local-runner: shutting down

</pre>`

Can anybody point out what wrong I am doing or a better way of doing it ? My goal is simply to enter card details through webdriverIO.

**Comment by christian-bromann** [comment]
Can you provide a reproducible example?

**Reply by arsalanator** [reply]
Thanks @christian-bromann  for asking for the reproducible example, I removed unnecessary stuff from it and made it for you. here it is https://github.com/soberservicesguy/stripeForWebdriver

**Comment by christian-bromann** [comment]
Here is the git diff to get this working:
```
diff --git a/test/pageobjects/page.js b/test/pageobjects/page.js
index 71f1207..8c90a85 100644
--- a/test/pageobjects/page.js
+++ b/test/pageobjects/page.js
@@ -1,6 +1,6 @@
 class Page {
-    open (path) {
-        return browser.url(`http://localhost:3000${path}`)
+    async open (path) {
+        await browser.url(`http://localhost:3000${path}`)
     }
 }
 
diff --git a/test/pageobjects/settings.page.js b/test/pageobjects/settings.page.js
index 5f41c01..3b11cb8 100644
--- a/test/pageobjects/settings.page.js
+++ b/test/pageobjects/settings.page.js
@@ -8,12 +8,9 @@ class SettingsPage extends Page {
      * define selectors using getter methods
      */
    
-    get stripeCardNumberInput () { 
-        let outerDiv = $('#stripeCheckout')
-        return outerDiv.$('.InputElement') 
-    } // $('[name="cardnumber"]')
-    get stripeCardExpiryInput () { return  } // $('[name="cardExpiry"]')
-    get stripeCardCVCInput () { return  } // $('[name="cvc"]')
+    get stripeCardNumberInput () { return $('input[name="cardnumber"]') } // $('[name="cardnumber"]')
+    get stripeCardExpiryInput () { return $('input[name="exp-date"]') }
+    get stripeCardCVCInput () { return $('input[name="cvc"]') }
     get subscribeButton () { return $('#subscribeButton') }
 
     /**
@@ -21,11 +18,13 @@ class SettingsPage extends Page {
      * e.g. to login using username and password
      */
     async waitForStripeCheckout () {
-        $('#stripeCheckout').waitForClickable({ timeout:5000 })
+        return $('#stripeCheckout').waitForClickable({ timeout:5000 })
         // $('#stripeCheckout').waitForExist({ timeout:5000 })
     }
 
     async enterCardDetailsAndSubmit (cardNumber, cardExpiry, cvc) {
+        const iframe = await $('iframe')
+        await browser.switchToFrame(iframe)
         await this.stripeCardNumberInput.setValue(cardNumber);
         await this.stripeCardExpiryInput.setValue(cardExpiry);
         await this.stripeCardCVCInput.setValue(cvc);
```

The problem is that the form elements are within an iframe and you need to switch to that one first to interact with these elements, see

```
+        const iframe = await $('iframe')
+        await browser.switchToFrame(iframe)
```
